### PR TITLE
[ML] Add early stopping DFA configuration parameter

### PR DESCRIPTION
**Pull request by valeriy42**
The PR adds `early_stopping_enabled` optional data frame analysis configuration parameter. The enhancement was already described in elastic/ml-cpp#1676 and so I mark it here as `non-issue`.

### CODE REVIEW

**Comment by valeriy42** [comment]
@szabosteve could you please take a look at the docs?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java [hunk 1]
```diff
@@ -214,6 +224,7 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalDouble(softTreeDepthTolerance);
         out.writeOptionalDouble(downsampleFactor);
         out.writeOptionalVInt(maxOptimizationRoundsPerHyperparameter);
+        out.writeOptionalBoolean(earlyStoppingEnabled);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java:227]: This needs to be protected by version guards, for example:

```
if (out.getVersion().onOrAfter(V_7_12_0)) {
    out.writeOptionalBoolean(earlyStoppingEnabled);
}
```

Whether you can get away with `V_7_12_0` from the start or whether you have to use `V_8_0_0` in the initial master branch PR, then change to `V_7_12_0` after backporting to 7.x depends on how good our BWC coverage is: if we have good coverage you'll need to start off with `V_8_0_0` and if we don't you can go straight to `V_7_12_0`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java [hunk 2]
```diff
@@ -150,6 +155,7 @@ public BoostedTreeParams(@Nullable Double lambda,
         softTreeDepthTolerance = in.readOptionalDouble();
         downsampleFactor = in.readOptionalDouble();
         maxOptimizationRoundsPerHyperparameter = in.readOptionalVInt();
+        earlyStoppingEnabled = in.readOptionalBoolean();
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java:158]: As for write, this needs a version guard, e.g.:

```
if (in.getVersion().onOrAfter(V_8_0_0)) {
    earlyStoppingEnabled = in.readOptionalBoolean();
} else {
    earlyStoppingEnabled = null;
}
```

**Review by droberts195 — COMMENTED** [review]
I think you also need to add a config index mapping for the new field in `x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml`

**Comment by dimitris-athanasiou** [comment]
Do we consider `early_stopping_enabled` to be a boosted tree param? If not, I would suggest we move it outside of `BoostedTreeParams` and straight into `Regression` and `Classification`.

**Comment by valeriy42** [comment]
> Do we consider `early_stopping_enabled` to be a boosted tree param? If not, I would suggest we move it outside of `BoostedTreeParams` and straight into `Regression` and `Classification`.

Good point. It is not restricted to boosted trees.

File: docs/reference/ml/ml-shared.asciidoc [hunk 3]
```diff
@@ -557,6 +557,14 @@ Values must be greater than zero and less than or equal to 1.
 By default, this value is calculated during hyperparameter optimization.
 end::dfas-downsample-factor[]
 
+tag::dfas-early-stopping-enabled[]
+Advanced configuration option.
+Specifies whether the training process should finish if it is not finding any
+better perfoming models. If disabled, the training process can take significantly
+longer and result in a better performing model purely by random chance.
```
Comment (szabosteve) [inline docs/reference/ml/ml-shared.asciidoc:564]: Nit. Please take or leave this suggestion. If users just skim the text, they might misunderstand the original sentence by a higher chance. E.g.: "and result in a better performing model (...)". This is why I suggested this alternative wording.

```suggestion
longer and the chance of finding a better performing model is unremarkable.
```

**Review by szabosteve — APPROVED** [review]
Thanks for writing the docs. I left one suggestion, please take it or leave it.
LGTM.

**Comment by valeriy42** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java (hunk 1 above)
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java]: Fixed :+1:

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java (hunk 2 above)
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/BoostedTreeParams.java]: I changed the type to `boolean`

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks good! Just some minor improvements in covering `null` in tests.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java [hunk 4]
```diff
@@ -88,6 +88,7 @@ private static Regression createRandom(BoostedTreeParams boostedTreeParams) {
         Long randomizeSeed = randomBoolean() ? null : randomLong();
         Regression.LossFunction lossFunction = randomBoolean() ? null : randomFrom(Regression.LossFunction.values());
         Double lossFunctionParameter = randomBoolean() ? null : randomDoubleBetween(0.0, Double.MAX_VALUE, false);
+        Boolean earlyStoppingEnabled = randomBoolean();
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java:91]: Fixed in 248111e5c6b

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java [hunk 5]
```diff
@@ -97,6 +97,7 @@ public static Classification createRandom() {
         Integer numTopClasses = randomBoolean() ? null : randomIntBetween(-1, 1000);
         Double trainingPercent = randomBoolean() ? null : randomDoubleBetween(0.0, 100.0, false);
         Long randomizeSeed = randomBoolean() ? null : randomLong();
+        Boolean earlyStoppingEnabled = randomBoolean();
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java]: Fixed in 248111e5c6b

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ml/dataframe/RegressionTests.java [hunk 6]
```diff
@@ -59,6 +59,7 @@ public static Regression randomRegression() {
             .setSoftTreeDepthTolerance(randomBoolean() ? null : randomDoubleBetween(0.01, Double.MAX_VALUE, true))
             .setDownsampleFactor(randomBoolean() ? null : randomDoubleBetween(0.0, 1.0, false))
             .setMaxOptimizationRoundsPerHyperparameter(randomBoolean() ? null : randomIntBetween(0, 20))
+            .setEarlyStoppingEnabled(randomBoolean())
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ml/dataframe/RegressionTests.java]: Fixed in 248111e5c6b

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ml/dataframe/ClassificationTests.java [hunk 7]
```diff
@@ -60,6 +60,7 @@ public static Classification randomClassification() {
             .setSoftTreeDepthTolerance(randomBoolean() ? null : randomDoubleBetween(0.01, Double.MAX_VALUE, true))
             .setDownsampleFactor(randomBoolean() ? null : randomDoubleBetween(0.0, 1.0, false))
             .setMaxOptimizationRoundsPerHyperparameter(randomBoolean() ? null : randomIntBetween(0, 20))
+            .setEarlyStoppingEnabled(randomBoolean())
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ml/dataframe/ClassificationTests.java]: Fixed in 248111e5c6b

**Comment by valeriy42** [comment]
Thank you for the review @dimitris-athanasiou . I added the `null` tests as you suggested. It would be great if you could take another look.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Comment by valeriy42** [comment]
@elasticmachine update branch

**Pull request merged**